File: gtci/00-warmup/cpp/7-sqrt.cpp
#include <iostream>

class Solution {
   public:
    int mySqrt(int x) {
        if (x < 2) return x;

        int left = 2;
        int right = x / 2;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            long long num = (long long)mid * mid;

            if (num < x) {
                left = mid + 1;
            } else if (num > x) {
                right = mid - 1;
            } else {
                return mid;
            }
        }

        return right;
    }
};

#include <cassert>

int main() {
    Solution sol;
    assert(sol.mySqrt(4) == 2);
    assert(sol.mySqrt(8) == 2);
    assert(sol.mySqrt(0) == 0);
    assert(sol.mySqrt(1) == 1);
    std::cout << "All test cases passed." << std::endl;
    return 0;
}

File: gtci/00-warmup/python3/7-sqrt.py
def mySqrt(x: int) -> int:
    if x < 2:
        return x

    left = 2
    right = x // 2

    while left <= right:
        mid = left + (right - left) // 2
        num = mid * mid

        if num < x:
            left = mid + 1
        elif num > x:
            right = mid - 1
        else:
            return mid

    return right


if __name__ == "__main__":
    assert mySqrt(4) == 2
    assert mySqrt(8) == 2
    assert mySqrt(0) == 0
    assert mySqrt(1) == 1
    print("All test cases passed.")
